sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3469,6 +3470,52 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
     }
   }
 
+  /**
+   * Rewrite [[UnresolvedWithColumns]] into a [[Project]].
+   *
+   * `withColumns` can both update an existing columns and add new columns.
+   */
+  object RewriteWithColumns extends Rule[LogicalPlan] {
+    override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+      _.containsPattern(UNRESOLVED_WITH_COLUMNS), ruleId) {
+      case UnresolvedWithColumns(projectList, child) if child.resolved =>
+        SchemaUtils.checkColumnNameDuplication(
+          projectList.map(_.name),
+          resolver)
+        val usedProjectListIndices = mutable.Set.empty[Int]
+        val outputProjectListBuilder = Seq.newBuilder[NamedExpression]
+        val lookup = {
```

Comment:
shall we simplify the code a bit? Basically we only need two list
```
val updatedProjectList = ...
val extraProjectList = ...
child.output.foreach { col =>
  if (need to replace) 
    updatedProjectList += newExpr
  else 
    updatedProjectList += col
    extraProjectList += newExpr
}
Project(updatedProjectList ++ extraProjectList, ...)
```

Reply:
Yea we still need the `caseSensitiveLookup`, but we don't need to keep track a set of indices.

Reply:
So, AFAIK, we still support duplicate names in projects. I can replace the bitmap by something else, but it will still be similar.

Reply:
nvm, I misread your code.